craft_application/services/provider.py

```diff
@@ -157,22 +161,69 @@ def get_provider(self, name: str | None = None) -> craft_providers.Provider:
 
         :param name: if set, uses the given provider name.
 
+        The provider is determined in the following order:
+        (1) use provider specified in the function argument,
+        (2) get the provider from the environment (CRAFT_BUILD_ENVIRONMENT),
```

Comment:
This code works, but it's bothering me a bit from a design point-of-view that this determination of which provider to use, if it's destructive or not, etc is now kind of all over the place.

How about this refactoring (I haven't thought through this too much I confess):

1) Put this "chosen provider" login in a separate utility function somewhere;
2) Have *both* decisions - about whether to run in destructive mode, and of which provider to use - be made by the command class. That class already answers whether to `run_managed()`, and I like the idea of the (apparently unused?) `provider_name()` method, which could call the utility function from 1)
3) Have the ProviderService obey whatever is passed to it as the provider name, but fail if `None` is passed. Failing from `None` is kind of an early breaker because if the command is not managed, the provider service should not have been instantiated at all and there's probably a bug in the code.

I think this might make our lives easier moving forward, because then we can look just at the command to see what's going to be used, instead of having to look at the command and also at the provider. What do y'all think?

Reply:
I like that change. Perhaps we can deprecate `run_managed()` and instead just have a `get_provider_name()` that returns `str | None`? Then in `Application.run()` we can simplify our logic a bit:

```
provider_name = command.get_provider_name(parsed_args)

...

self._configure_services(platform, build_for, provider_name)

...

if provider_name and not self.is_managed():
    self.run_managed(platform, build_for)
    return_code = 0
else:
    if not self.is_managed():
        craft_cli.emit.debug(f"Running {self.app.name} {command.name} on host")
    return_code = dispatcher.run() or 0
```

This is already a breaking change, so it's not an issue to break it further.

Reply:
I don't know, I personally don't like that `get_provider_name()` returning `None` means a destructive build. It seems to conflate two things into one

Reply:
We could replace `get_provider_name()` for something more generic
```
def get_build_environment() -> str:
  """Returns "lxd", "multipass", or "host"."""
```

and retain `run_managed()` but simplified to:
```
def run_managed():
  return self.get_build_environment() != "host"
```